src/ShallowWaterExplicit.jl

```diff
@@ -0,0 +1,171 @@
+function shallow_water_explicit_time_step!(model, dΩ, dω, f, g, h₁, u₁, hₘ, uₘ, hₚ, uₚ, RTMMchol, L2MMchol, dt, leap_frog, τ, Q, V, R, S, h₂, u₂, ϕ, F)
+  # energetically balanced explicit second order shallow water solver
+  # reference: eqns (21-24) of
+  # https://github.com/BOM-Monash-Collaborations/articles/blob/main/energetically_balanced_time_integration/EnergeticallyBalancedTimeIntegration_SW.tex
+  #
+  # f          : coriolis force (field)
+  # g          : gravity (constant)
+  # h₁         : fluid depth at current time level
+  # u₁         : fluid velocity at current time level
+  # hₘ         : fluid depth at previous time level (for leap-frogging the first step)
+  # uₘ         : fluid velocity at previous time level (for leap-frogging the first step)
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # leap_frog  : do leap frog time integration for the first step (boolean)
+  # dΩ         : measure of the elements
+
+  n = get_normal_vector(model)
+  # explicit step for provisional velocity, uₚ
+  dt1 = dt
+  if leap_frog
+    dt1 = 2.0*dt
+  end
+
+  # 1.1: the mass flux
+  b₁(v)  = ∫(v⋅u₁*h₁)*dΩ
+  rhs1   = assemble_vector(b₁, V)
+  copy!(get_free_dof_values(F), rhs1)
+  ldiv!(RTMMchol, get_free_dof_values(F))
+  # 1.2: the bernoulli function
+  b₂(q)  = ∫(q*(0.5*u₁⋅u₁ + g*h₁))*dΩ
+  rhs2   = assemble_vector(b₂, Q)
+  copy!(get_free_dof_values(ϕ), rhs2)
+  ldiv!(L2MMchol, get_free_dof_values(ϕ))
+  # 1.3: the potential vorticity
+  a₁(r,s) = ∫(s*h₁*r)dΩ
+  c₁(s)   = ∫(perp(n,∇(s))⋅(u₁) + s*f)dΩ
+  H1h     = assemble_matrix(a₁, R, S)
```

Comment:
Although correct, this is not the most efficient way to assemble this problem. If you separate the assembly of the matrix and the right hand side, you will need two different loops over the mesh (once for the matrix and another for the vector). Better to fuse both loops in a single one. To this end,

(1) Remove `assemble_matrix` and `assemble_vector` calls.
(2) Replace `AffineFEOperator(R, S, H1h, rhs_q₁)` by `AffineFEOperator(a1, b1, R, S)`

Does this make sense?

Reply:
Yes, I didn't know the difference between these two. Will do, thanks!